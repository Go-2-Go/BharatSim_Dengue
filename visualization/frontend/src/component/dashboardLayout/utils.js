import React from 'react';
import renderChart from '../charts/renderChart';

export function renderElement(element) {
  return (
    <div key={element.i} data-grid={element} data-testid={element.i}>
      <div style={{ height: '100%', width: '100%', boxSizing: 'border-box', padding: '10px' }}>
        {renderChart(element.chartType, { config: element.config })}
      </div>
    </div>
  );
}

export function getNewWidgetLayout(numberOfWidgetAdded, cols, count) {
  return createLayout({
    id: `widget-${count}`,
    xPosition: (numberOfWidgetAdded * 2) % cols,
    yPosition: Infinity, // puts it at the bottom
    width: 2,
    height: 2,
  });
}

function createLayout({ id, xPosition, yPosition, width, height }) {
  return {
    i: id,
    x: xPosition,
    y: yPosition,
    w: width,
    h: height,
  };
}
