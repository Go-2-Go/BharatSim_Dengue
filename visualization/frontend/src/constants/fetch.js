export const httpMethods = {
  GET: 'get',
  POST: 'post',
};

export const contentTypes = {
  FILE: 'multipart/form-data',
};
