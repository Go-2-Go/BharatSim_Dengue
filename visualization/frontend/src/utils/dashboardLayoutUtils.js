/* eslint-disable react/prop-types */

import React from 'react';
import renderChart from '../component/charts/renderChart';

export function renderElement({ layout, chartType, config }) {
  return (
    <div key={layout.i} data-grid={layout} data-testid={layout.i}>
      <div style={{ height: '100%', width: '100%', boxSizing: 'border-box', padding: '10px' }}>
        {renderChart(chartType, { config })}
      </div>
    </div>
  );
}

export function getNewWidgetLayout(numberOfWidgetAdded, cols, count) {
  return createLayout({
    id: `widget-${count}`,
    xPosition: (numberOfWidgetAdded * 2) % cols,
    yPosition: Infinity, // puts it at the bottom
    width: 2,
    height: 2,
  });
}

function createLayout({ id, xPosition, yPosition, width, height }) {
  return {
    i: id,
    x: xPosition,
    y: yPosition,
    w: width,
    h: height,
  };
}
