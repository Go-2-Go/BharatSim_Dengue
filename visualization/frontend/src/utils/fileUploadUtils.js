const fileUploadedStatus = {
  LOADING: 'loading',
  SUCCESS: 'success',
  ERROR: 'error',
};

function getStatusAndMessageFor(fileUploadStatus, fileName) {
  return {
    [fileUploadedStatus.ERROR]: {
      status: fileUploadedStatus.ERROR,
      message: `Error occurred while unloading ${fileName}`,
    },
    [fileUploadedStatus.SUCCESS]: {
      status: fileUploadedStatus.SUCCESS,
      message: `${fileName} successfully uploaded`,
    },
    [fileUploadedStatus.LOADING]: {
      status: fileUploadedStatus.LOADING,
      message: `uploading ${fileName}`,
    },
  }[fileUploadStatus];
}

export { fileUploadedStatus, getStatusAndMessageFor };
