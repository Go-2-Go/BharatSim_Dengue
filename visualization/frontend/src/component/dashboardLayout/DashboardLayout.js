import React, { useState } from 'react';
import PropTypes from 'prop-types';

import { Box, Button, withStyles } from '@material-ui/core';
import ReactGridLayout, { WidthProvider } from 'react-grid-layout';

import 'react-grid-layout/css/styles.css';
import 'react-resizable/css/styles.css';
import styles from './dashboardLayoutCss';

import { renderElement, getNewWidgetLayout } from './utils';
import labels from '../../constants/labels';
import ChartConfigModal from '../chartConfigModal/ChartConfigModal';
import useModal from '../../hook/useModal';
import FileUpload from '../../uiComponent/FileUpload';
import { chartTypes } from '../../constants/charts';

const GridLayout = WidthProvider(ReactGridLayout);
const cols = 12;

const DashboardLayout = ({ classes }) => {
  const [widgets, setWidgets] = useState([]);
  const [count, setCount] = useState(0);
  const [layout, setLayout] = useState();
  const [chartType, setChartType] = useState();

  const { isOpen, closeModal, openModal } = useModal();

  const addItem = (config) => {
    setWidgets((prevWidgets) => {
      return prevWidgets.concat({
        ...getNewWidgetLayout(prevWidgets.length, cols, count),
        config,
        chartType,
      });
    });
    setCount((prevCount) => prevCount + 1);
  };

  const handleModalOk = (config) => {
    addItem(config);
    closeModal();
  };

  const onLayoutChange = (changedLayout) => {
    setLayout(changedLayout);
  };
  const oneChartClick = (selectedChartType) => {
    openModal();
    setChartType(selectedChartType);
  };

  return (
    <Box pl={10} pt={10} pr={10}>
      <Box pb={2}>
        <FileUpload />
      </Box>
      <Box pb={2}>
        <Button
          onClick={() => oneChartClick(chartTypes.BAR_CHART)}
          variant="contained"
          color="primary"
          className={classes.buttonRoot}
        >
          {labels.dashboardLayout.Bar_CHART}
        </Button>
        <Button
          onClick={() => oneChartClick(chartTypes.LINE_CHART)}
          variant="contained"
          color="primary"
          className={classes.buttonRoot}
        >
          {labels.dashboardLayout.LINE_CHART}
        </Button>
      </Box>
      <GridLayout
        layout={layout}
        onLayoutChange={onLayoutChange}
        className={classes.reactGridLayout}
      >
        {widgets.map((item) => {
          return renderElement(item);
        })}
      </GridLayout>
      {isOpen && (
        <ChartConfigModal
          onCancel={closeModal}
          onOk={handleModalOk}
          open={isOpen}
          chartType={chartType}
        />
      )}
    </Box>
  );
};

DashboardLayout.propTypes = {
  classes: PropTypes.shape({
    buttonRoot: PropTypes.string.isRequired,
    reactGridLayout: PropTypes.string.isRequired,
  }).isRequired,
};

export default withStyles(styles)(DashboardLayout);
