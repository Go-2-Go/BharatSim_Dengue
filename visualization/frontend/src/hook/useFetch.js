import { useEffect, useState } from 'react';
import fetch from '../utils/fetch';
import { httpMethods } from '../constants/httpMethods';

const useFetch = ({ url, method = httpMethods.GET, headers, data, query }) => {
  const [responseData, setResponseData] = useState();

  useEffect(() => {
    const fetchData = async () => {
      const resData = await fetch({ url, method, headers, data, query });
      setResponseData(resData);
    };
    fetchData();
  }, []);

  return responseData;
};

export default useFetch;
