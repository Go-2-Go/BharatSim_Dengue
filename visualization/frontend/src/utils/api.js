import { contentTypes, httpMethods } from '../constants/fetch';
import { url as serviceURL } from './url';
import { fetchData, uploadData } from './fetch';

function headerBuilder({ contentType }) {
  return {
    'content-type': contentType,
  };
}

function formDataBuilder(data) {
  const formData = new FormData();
  data.forEach(({ name, value }) => formData.append(name, value));
  return formData;
}

const api = {
  saveDashboard: async ({ widgets, layout, dashboardId, name, count }) => {
    return uploadData({
      url: serviceURL.DASHBOARD_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      data: JSON.stringify({
        dashboardData: { widgets, layout, dashboardId, name, count },
      }),
    });
  },

  getAllDashBoard: async () => fetchData({ url: serviceURL.DASHBOARD_URL }),

  uploadFileAndSchema: async ({ file, schema }) =>
    uploadData({
      url: serviceURL.DATA_SOURCES,
      data: formDataBuilder([
        { name: 'datafile', value: file },
        { name: 'schema', value: JSON.stringify(schema) },
      ]),
      headers: headerBuilder({ contentType: contentTypes.FILE }),
    }),

  getCsvHeaders: async (dataSourceId) => fetchData({ url: serviceURL.getHeaderUrl(dataSourceId) }),

  getDatasources: async () => fetchData({ url: serviceURL.DATA_SOURCES }),

  getData: async (datasource, query) =>
    fetchData({ url: serviceURL.getDataUrl(datasource), query }),

  saveProject: async ({ id, ...data }) => {
    const requestObject = {
      url: serviceURL.PROJECT_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
    };
    if (id) {
      return uploadData({
        ...requestObject,
        data: JSON.stringify({ projectData: { ...data, id } }),
        method: httpMethods.PUT,
      });
    }
    return uploadData({
      ...requestObject,
      data: JSON.stringify({ projectData: { ...data } }),
      method: httpMethods.POST,
    });
  },

  getProjects: async () => {
    return fetchData({
      url: serviceURL.PROJECT_URL,
    });
  },
  getProject: async (id) => {
    return fetchData({
      url: serviceURL.getProjectUrl(id),
    });
  },
};

export { api };
