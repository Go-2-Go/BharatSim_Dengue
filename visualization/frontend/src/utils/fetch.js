import axios from 'axios';
import { httpMethods } from '../constants/fetch';

let shouldShowError;

function commonErrorHandler(err) {
  shouldShowError(true);
}

function initApiConfig({ setError }) {
  shouldShowError = setError;
}

function fetchData({ url, method = httpMethods.GET, headers, data, query, customErrorHandler }) {
  return axios({ url, method, headers, data, params: query })
    .then((res) => res.data)
    .catch((err) => {
      if (customErrorHandler) {
        customErrorHandler(err);
        return;
      }
      commonErrorHandler(err);
    });
}

function uploadData({ url, headers, data, query, customErrorHandler }) {
  return fetchData({
    url,
    method: httpMethods.POST,
    headers,
    data,
    params: query,
    customErrorHandler,
  });
}

export { fetchData, uploadData, initApiConfig };
