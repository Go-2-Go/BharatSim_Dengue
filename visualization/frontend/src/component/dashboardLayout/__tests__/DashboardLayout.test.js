import React from 'react';
import { act, fireEvent, render } from '@testing-library/react';
import DashboardLayout from '../DashboardLayout';
import useFetch from '../../../hook/useFetch';
import * as fetch from '../../../utils/fetch';

jest.spyOn(fetch, 'uploadData');
jest.spyOn(fetch, 'fetchData');

jest.mock('../../chartConfigModal/ChartConfigModal', () => ({ open, onCancel, onOk }) => (
  <>
    <span>
      Modal Open
      {`${open.toString()}`}
    </span>
    <button type="button" onClick={() => onOk('config')}>
      Ok
    </button>
    <button type="button" onClick={onCancel}>
      Cancel
    </button>
  </>
));

jest.mock('../../fileUpload/FileUpload', () => (props) => (
  <div>
    File Upload
    {/* eslint-disable-next-line no-undef */}
    <span>{mockPropsCapture(props)}</span>
  </div>
));

jest.mock('../../charts/barChart/BarChart', () => (props) => (
  <div>
    Bar Chart
    {/* eslint-disable-next-line no-undef */}
    <span>{mockPropsCapture(props)}</span>
  </div>
));

jest.mock('../../charts/lineChart/LineChart', () => (props) => (
  <div>
    Line Chart
    {/* eslint-disable-next-line no-undef */}
    <span>{mockPropsCapture(props)}</span>
  </div>
));

jest.mock('../../../hook/useFetch');

describe('<DashboardLayout />', () => {
  beforeEach(() => {
    useFetch.mockReturnValue({ headers: ['x-header', 'y-header'] });
    fetch.uploadData.mockResolvedValue({ dashboardId: 'id' });
    fetch.fetchData.mockResolvedValue({ dashboards: [] });
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match a snapshot for <DashboardLayout />', () => {
    const { container } = render(<DashboardLayout />);

    expect(container).toMatchSnapshot();
  });
  it('should match a snapshot for <DashboardLayout /> if dashboard is loaded with data', async () => {
    const data = {
      name: 'dashboard1',
      widgets: [
        {
          layout: { i: 'widget-0', x: 0, y: null, w: 2, h: 2 },
          config: 'config',
          chartType: 'barChart',
        },
      ],
      dashboardId: null,
      layout: [{ w: 2, h: 2, x: 0, y: 0, i: 'widget-0', moved: false, static: false }],
    };
    fetch.fetchData.mockResolvedValue({ dashboards: [data] });

    const { container, findByTestId } = render(<DashboardLayout />);

    await findByTestId('widget-0');

    expect(container).toMatchSnapshot();
  });

  it('should open modal on click of cancel button ', () => {
    const { getByText } = render(<DashboardLayout />);

    const addWidgetButton = getByText(/Line chart/i);
    fireEvent.click(addWidgetButton);

    expect(getByText(/Modal Open/i)).toHaveTextContent(/true/i);
  });

  it('should closed modal on click of cancel button ', async () => {
    const { getByText, queryByText } = render(<DashboardLayout />);

    const addWidgetButton = getByText(/Line Chart/i);
    fireEvent.click(addWidgetButton);

    const cancelButton = getByText(/Cancel/i);
    fireEvent.click(cancelButton);

    expect(await queryByText(/Modal Open/i)).toBeNull();
  });

  it('should add new widget with line chart', () => {
    const { getByText, getByTestId } = render(<DashboardLayout />);

    const addWidgetButton = getByText(/Line Chart/i);
    fireEvent.click(addWidgetButton);

    const okButton = getByText(/Ok/i);
    fireEvent.click(okButton);

    const widget = getByTestId('widget-0');

    expect(widget).toBeInTheDocument();
    expect(widget).toMatchSnapshot();
  });
  it('should add new widget with bar chart', () => {
    const { getByText, getByTestId } = render(<DashboardLayout />);

    const addWidgetButton = getByText(/Bar Chart/i);
    fireEvent.click(addWidgetButton);

    const okButton = getByText(/Ok/i);
    fireEvent.click(okButton);

    const widget = getByTestId('widget-0');

    expect(widget).toBeInTheDocument();
    expect(widget).toMatchSnapshot();
  });
  it('should save dashboard on click of save dashboard button with empty widgets', async () => {
    const { getByText } = render(<DashboardLayout />);
    const saveDashboard = getByText(/Save Dashboard/i);
    const requestObject = {
      data: JSON.stringify({
        dashboardData: {
          widgets: [],
          layout: [],
          dashboardId: null,
          name: 'dashboard1',
          count: 0,
        },
      }),
      headers: { 'content-type': 'application/json' },
      url: '/api/dashboard',
    };

    await act(async () => {
      fireEvent.click(saveDashboard);
    });

    expect(fetch.uploadData).toHaveBeenCalledWith(requestObject);
  });
  it('should save dashboard with widgets on click of save dashboard button ', async () => {
    const { getByText } = render(<DashboardLayout />);
    const saveDashboard = getByText(/Save Dashboard/i);

    const requestObject = {
      data: JSON.stringify({
        dashboardData: {
          widgets: [
            {
              config: 'config',
              chartType: 'barChart',
              layout: { i: 'widget-0', x: 0, y: null, w: 2, h: 2 },
            },
          ],
          layout: [{ w: 2, h: 2, x: 0, y: 0, i: 'widget-0', moved: false, static: false }],
          dashboardId: null,
          name: 'dashboard1',
          count: 1,
        },
      }),
      headers: { 'content-type': 'application/json' },
      url: '/api/dashboard',
    };

    const addWidgetButton = getByText(/Bar Chart/i);
    fireEvent.click(addWidgetButton);

    const okButton = getByText(/Ok/i);
    fireEvent.click(okButton);

    await act(async () => {
      fireEvent.click(saveDashboard);
    });
    expect(fetch.uploadData).toHaveBeenCalledWith(requestObject);
  });
});
