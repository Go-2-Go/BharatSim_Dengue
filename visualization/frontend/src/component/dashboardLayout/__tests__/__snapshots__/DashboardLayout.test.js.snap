// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DashboardLayout /> should add new widget with bar chart 1`] = `
<div
  class="react-grid-item react-draggable cssTransforms react-resizable"
  data-grid="[object Object]"
  data-testid="widget-0"
  style="transform: translate(10px,10px); width: -12px; height: 310px; position: absolute;"
>
  <div
    style="height: 100%; width: 100%; box-sizing: border-box; padding: 10px;"
  >
    <div>
      Bar Chart
      <span>
        config = [object Object]
      </span>
    </div>
  </div>
  <span
    class="react-resizable-handle react-resizable-handle-se"
  />
</div>
`;

exports[`<DashboardLayout /> should add new widget with line chart 1`] = `
<div
  class="react-grid-item react-draggable cssTransforms react-resizable"
  data-grid="[object Object]"
  data-testid="widget-0"
  style="transform: translate(10px,10px); width: -12px; height: 310px; position: absolute;"
>
  <div
    style="height: 100%; width: 100%; box-sizing: border-box; padding: 10px;"
  >
    <div>
      Line Chart
      <span>
        config = [object Object]
      </span>
    </div>
  </div>
  <span
    class="react-resizable-handle react-resizable-handle-se"
  />
</div>
`;

exports[`<DashboardLayout /> should match a snapshot for <DashboardLayout /> 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-3"
  >
    <div
      class="MuiBox-root MuiBox-root-4"
    >
      <div>
        File Upload
        <span />
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-5"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained DashboardLayout-buttonRoot-1 MuiButton-containedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Bar Chart
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained DashboardLayout-buttonRoot-1 MuiButton-containedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Line Chart
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="react-grid-layout DashboardLayout-reactGridLayout-2"
      style="height: 10px;"
    />
  </div>
</div>
`;
